Fix MemberServices endpoint to update members data

rodrei opened the pull request.

Tuuleh commented on app/services/member_services_member_service.rb:
```diff
@@ -1,28 +0,0 @@
-# frozen_string_literal: true
-
-# It's a member service... for member services!
-class MemberServicesMemberService
```
Awww I was so proud of my punny class name.

Tuuleh commented on spec/requests/api/member_services/members_spec.rb:
```diff
@@ -4,110 +4,96 @@
+        }
+      end
 
-        context 'when the update fails' do
```
Why did you remove error handling from when an update just fails?

rodrei replied:
It's still there, but rewrote it to actually hit the service instead of mocking the response.

Tuuleh commented on app/services/member_services_member_service.rb:
```diff
@@ -1,28 +0,0 @@
-    if @member.update_attributes(@params)
-      return true
-    else
-      @errors = ["Updating member details failed for #{@member.email}."]
```
I would keep this in. It's nice to get a specific message when there is an issue updating attributes, rather than getting a random 500 that'll take a good while to debug.

rodrei replied:
If there's a problem with the update, the api will return a 422. If you're debugging this case all the information is already in the standard logs (response status, params) , so logging this is redundant.

Tuuleh commented on app/services/member_services/member_updater.rb:
```diff
@@ -0,0 +1,23 @@
+    end
+
+    def run
+      @member = Member.find_by_email(@email)
```
So I guess this PR is not to make it possible to update an email...?

rodrei replied:
`email` is the identifier field used to find a member.
`params` is meant to include the values to be updated, it can include `email`.

Tuuleh replied:
Ah I see, there are two email keys. One top level key for searching the member, and one in the `member` hash. Gotcha.

Tuuleh commented on app/controllers/api/member_services_controller.rb:
```diff
@@ -24,12 +24,14 @@ def gocardless_customers
-    if @member_service.update
-      render 'api/member_services/member', status: 200
+    email = params.require(:email)
+    member_params = params.require(:member).permit(:email, :first_name, :last_name, :country, :postal)
```
I was debugging the lambda invocations from the DynamoDB table and while doing that I ran into this error:
```Jan 17 02:57:20 env-staging production.log:  I, [2018-01-17T10:57:20.811560 #5]  INFO -- : method=PUT path=/api/member_services/members format=html controller=Api::MemberServicesController action=update_member status=400 error='ActionController::ParameterMissing: param is missing or the value is empty: member' duration=0.56 view=0.00 db=0.00 params={"email"=>"tuuli@sumofus.org", "first_name"=>"Tuuli", "last_name"=>"Pollanen", "country"=>"Slovenia", "member_service"=>{"email"=>"tuuli@sumofus.org", "first_name"=>"Tuuli", "last_name"=>"Pollanen", "country"=>"Slovenia"}} time=2018-01-17 10:57:20 +0000 exception=["ActionController::ParameterMissing", "param is missing or the value is empty: member"]
```

rodrei replied:
Sorry! I probably should've deployed development back to staging after finishing with my work.

The behaviour you see is the expected one. `email` is a required param so if it's missing a 400 Bad Request is returned, and the error gets logged.

Tuuleh replied:
It was missing the `member` param but I see it's been added in your front end PR :+1:

Tuuleh commented:
I'm not totally sure why not just e.g. pass the initial value of the email address in an `old_email` or `search_by_email` field and pass that in `search_by_email`. No other changes are really required.

rodrei commented:
@Tuuleh Initially I was just going to do something like what you suggested, but then I saw there were a few things that could be easily improved/fixed, so I just went ahead and did it:

* I edited the format of the params received. Instead of sending all the params at the top level, I changed it to the standard of putting them within a `member` key. 
* I renamed the Service to a more meaningful name. 
* I simplified the service by extracting the knowledge about http status out of it. The service should only handle updating a member, how to manage http status is the controller's job.

Tuuleh commented on circle.yml:
```diff
@@ -36,7 +36,7 @@ deployment:
       - ./bin/deploy.sh $CIRCLE_SHA1 'champaign' 'env-production-1' 'champaign-assets-production' 'logs3.papertrailapp.com:44107' 'actions.sumofus.org'
   staging:
-    branch:  'development'
+    branch:  'email-update'
```
Remember to change the branch back to development.

Tuuleh commented:
This looks good to go, Rodri.

On Friday you mentioned that you often find yourself to be the only person doing refactoring when fixing things / adding logic into our existing code base. Is this the kind of refactoring you had in mind then?

Tuuleh commented:
+ merge pending the other related PR's getting a review and changing deploy branch back to development

rodrei commented:
> On Friday you mentioned that you often find yourself to be the only person doing refactoring when fixing things / adding logic into our existing code base. Is this the kind of refactoring you had in mind then?

I guess you could say that, yeah.

> merge pending the other related PR's getting a review and changing deploy branch back to development

Thanks for the reminder
